File: examples/primitives/address/sorting-filtering.ts
/**
 * Address Sorting and Filtering Example
 *
 * Demonstrates:
 * - Sorting addresses lexicographically
 * - Filtering addresses (zero addresses, ranges, sets)
 * - Deduplication and uniqueness
 * - Comparison operations
 * - Working with address collections
 */

import { Address } from "../../../src/primitives/Address/index.js";

console.log("=== Address Sorting and Filtering ===\n");

// 1. Sorting addresses
console.log("1. Sorting Addresses\n");

const unsorted = [
	Address.fromHex("0xffffffffffffffffffffffffffffffffffffffff"),
	Address.fromHex("0x0000000000000000000000000000000000000001"),
	Address.fromHex("0x742d35Cc6634C0532925a3b844Bc9e7595f51e3e"),
	Address.fromHex("0x0000000000000000000000000000000000000000"),
	Address.fromHex("0xa0Cf798816D4b9b9866b5330EEa46a18382f251e"),
];

console.log("Unsorted:");
unsorted.forEach((addr) => console.log(`  ${addr.toChecksummed()}`));
console.log();

// Sort ascending
const ascending = [...unsorted].sort((a, b) => a.compare(b));
console.log("Sorted (ascending):");
ascending.forEach((addr) => console.log(`  ${addr.toChecksummed()}`));
console.log();

// Sort descending
const descending = [...unsorted].sort((a, b) => b.compare(a));
console.log("Sorted (descending):");
descending.forEach((addr) => console.log(`  ${addr.toChecksummed()}`));
console.log();

// 2. Filtering zero addresses
console.log("2. Filtering Zero Addresses\n");

const addresses = [
	Address.zero(),
	Address.fromHex("0x742d35Cc6634C0532925a3b844Bc9e7595f51e3e"),
	new Address(0n),
	Address.fromHex("0xa0Cf798816D4b9b9866b5330EEa46a18382f251e"),
	Address.fromHex("0x0000000000000000000000000000000000000000"),
];

console.log("All addresses:");
addresses.forEach((addr) => {
	console.log(`  ${addr.toChecksummed()} ${addr.isZero() ? "(zero)" : ""}`);
});
console.log();

const nonZero = addresses.filter((addr) => !addr.isZero());
console.log(`Non-zero addresses (${nonZero.length}):)`);
nonZero.forEach((addr) => console.log(`  ${addr.toChecksummed()}`));
console.log();

// 3. Range filtering
console.log("3. Range Filtering\n");

const allAddresses = [
	new Address(10n),
	new Address(50n),
	new Address(100n),
	new Address(200n),
	new Address(500n),
];

const min = new Address(50n);
const max = new Address(200n);

console.log(`Range: ${min.toHex()} to ${max.toHex()}`);
console.log();

const inRange = allAddresses.filter(
	(addr) => !addr.lessThan(min) && !addr.greaterThan(max),
);

console.log("Addresses in range:");
inRange.forEach((addr) => console.log(`  ${addr.toHex()}`));
console.log();

// 4. Deduplication
console.log("4. Deduplication\n");

const withDuplicates = [
	Address.fromHex("0x742d35Cc6634C0532925a3b844Bc9e7595f51e3e"),
	Address.fromHex("0xa0Cf798816D4b9b9866b5330EEa46a18382f251e"),
	Address.fromHex("0x742d35cc6634c0532925a3b844bc9e7595f51e3e"), // Duplicate (different case)
	Address.fromHex("0xa0Cf798816D4b9b9866b5330EEa46a18382f251e"), // Duplicate (exact)
	Address.fromHex("0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266"),
];

console.log(`Original (${withDuplicates.length} addresses):`);
withDuplicates.forEach((addr) => console.log(`  ${addr.toChecksummed()}`));
console.log();

// Method 1: Using findIndex
const unique1 = withDuplicates.filter(
	(addr, i, arr) => arr.findIndex((a) => a.equals(addr)) === i,
);

console.log(`Deduplicated (${unique1.length} addresses):`);
unique1.forEach((addr) => console.log(`  ${addr.toChecksummed()}`));
console.log();

// Method 2: Using Set with hex representation
const hexSet = new Set<string>();
const unique2 = withDuplicates.filter((addr) => {
	const hex = addr.toHex();
	if (hexSet.has(hex)) return false;
	hexSet.add(hex);
	return true;
});

console.log(`Deduplicated with Set (${unique2.length} addresses):`);
unique2.forEach((addr) => console.log(`  ${addr.toChecksummed()}`));
console.log();

// 5. Finding specific addresses
console.log("5. Finding Specific Addresses\n");

const addressList = [
	Address.fromHex("0x742d35Cc6634C0532925a3b844Bc9e7595f51e3e"),
	Address.fromHex("0xa0Cf798816D4b9b9866b5330EEa46a18382f251e"),
	Address.fromHex("0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266"),
];

const target = Address.fromHex("0xa0Cf798816D4b9b9866b5330EEa46a18382f251e");

// Find index
const index = addressList.findIndex((addr) => addr.equals(target));
console.log(`Target: ${target.toChecksummed()}`);
console.log(`Found at index: ${index}`);
console.log();

// Check if exists
const exists = addressList.some((addr) => addr.equals(target));
console.log(`Exists in list: ${exists}`);
console.log();

// 6. Grouping by prefix
console.log("6. Grouping by Prefix\n");

const mixedAddresses = [
	Address.fromHex("0x0000000000000000000000000000000000000001"),
	Address.fromHex("0x0000000000000000000000000000000000000002"),
	Address.fromHex("0x1111111111111111111111111111111111111111"),
	Address.fromHex("0x1111111111111111111111111111111111111112"),
	Address.fromHex("0xffffffffffffffffffffffffffffffffffffffff"),
];

const grouped = new Map<number, Address[]>();

for (const addr of mixedAddresses) {
	const firstByte = addr[0]; // First byte determines group
	if (!grouped.has(firstByte)) {
		grouped.set(firstByte, []);
	}
	grouped.get(firstByte)!.push(addr);
}

console.log("Grouped by first byte:");
for (const [byte, addrs] of grouped.entries()) {
	console.log(
		`  0x${byte.toString(16).padStart(2, "0")}: ${addrs.length} addresses`,
	);
	addrs.forEach((addr) => console.log(`    ${addr.toChecksummed()}`));
}
console.log();

// 7. Top N addresses
console.log("7. Top N Addresses (Largest Values)\n");

const manyAddresses = [
	new Address(100n),
	new Address(500n),
	new Address(50n),
	new Address(1000n),
	new Address(200n),
	new Address(750n),
];

const topN = 3;
const largest = [...manyAddresses]
	.sort((a, b) => b.compare(a)) // Descending
	.slice(0, topN);

console.log(`Top ${topN} addresses:`);
largest.forEach((addr, i) => {
	console.log(`  ${i + 1}. ${addr.toHex()} (${addr.toU256()})`);
});
console.log();

// 8. Building address index/map
console.log("8. Building Address Index\n");

interface AddressMetadata {
	label: string;
	category: string;
}

const addressIndex = new Map<string, AddressMetadata>();

function addToIndex(addr: Address, meta: AddressMetadata) {
	addressIndex.set(addr.toHex(), meta);
}

function lookupAddress(addr: Address): AddressMetadata | undefined {
	return addressIndex.get(addr.toHex());
}

// Add addresses
addToIndex(Address.fromHex("0x742d35Cc6634C0532925a3b844Bc9e7595f51e3e"), {
	label: "Treasury",
	category: "Protocol",
});
addToIndex(Address.fromHex("0xa0Cf798816D4b9b9866b5330EEa46a18382f251e"), {
	label: "Deployer",
	category: "Admin",
});

// Lookup
const lookupAddr = Address.fromHex(
	"0x742d35Cc6634C0532925a3b844Bc9e7595f51e3e",
);
const metadata = lookupAddress(lookupAddr);
console.log(`Address: ${lookupAddr.toChecksummed()}`);
console.log(`Metadata: ${metadata ? JSON.stringify(metadata) : "Not found"}`);
console.log();

// 9. Performance note
console.log("9. Performance Note\n");
console.log("For optimal performance when working with collections:");
console.log("- Use Map<string, T> with address.toHex() as key for lookups");
console.log("- Use address.compare() for sorting (returns -1/0/1)");
console.log("- Use address.equals() for exact matching");
console.log("- Convert to hex once and reuse for multiple operations");

File: packages/voltaire-zig/examples/primitives/address/sorting-filtering.zig
const std = @import("std");
const primitives = @import("primitives");
const Address = primitives.address.Address;

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const a = gpa.allocator();

    std.debug.print("\n=== Address Sorting and Filtering ===\n\n", .{});

    // 1. Sorting addresses
    std.debug.print("1. Sorting Addresses\n\n", .{});

    var unsorted = [_]Address{
        try Address.fromHex("0xffffffffffffffffffffffffffffffffffffffff"),
        try Address.fromHex("0x0000000000000000000000000000000000000001"),
        try Address.fromHex("0x742d35Cc6634C0532925a3b844Bc9e7595f51e3e"),
        try Address.fromHex("0x0000000000000000000000000000000000000000"),
        try Address.fromHex("0xa0Cf798816D4b9b9866b5330EEa46a18382f251e"),
    };

    std.debug.print("Unsorted:\n", .{});
    for (unsorted) |addr| {
        const hex = Address.toChecksummed(addr);
        std.debug.print("  {s}\n", .{&hex});
    }
    std.debug.print("\n", .{});

    // Sort ascending
    std.mem.sort(Address, &unsorted, {}, struct {
        fn lessThan(_: void, lhs: Address, rhs: Address) bool {
            return Address.compare(lhs, rhs) < 0;
        }
    }.lessThan);

    std.debug.print("Sorted (ascending):\n", .{});
    for (unsorted) |addr| {
        const hex = Address.toChecksummed(addr);
        std.debug.print("  {s}\n", .{&hex});
    }
    std.debug.print("\n", .{});

    // 2. Filtering zero addresses
    std.debug.print("2. Filtering Zero Addresses\n\n", .{});

    const addresses = [_]Address{
        Address.zero(),
        try Address.fromHex("0x742d35Cc6634C0532925a3b844Bc9e7595f51e3e"),
        Address.fromU256(0),
        try Address.fromHex("0xa0Cf798816D4b9b9866b5330EEa46a18382f251e"),
        try Address.fromHex("0x0000000000000000000000000000000000000000"),
    };

    std.debug.print("All addresses:\n", .{});
    for (addresses) |addr| {
        const hex = Address.toChecksummed(addr);
        const is_zero = if (Address.isZero(addr)) " (zero)" else "";
        std.debug.print("  {s}{s}\n", .{ &hex, is_zero });
    }
    std.debug.print("\n", .{});

    var non_zero_list = std.ArrayList(Address).init(a);
    defer non_zero_list.deinit();

    for (addresses) |addr| {
        if (!Address.isZero(addr)) {
            try non_zero_list.append(addr);
        }
    }

    std.debug.print("Non-zero addresses ({}):\n", .{non_zero_list.items.len});
    for (non_zero_list.items) |addr| {
        const hex = Address.toChecksummed(addr);
        std.debug.print("  {s}\n", .{&hex});
    }
    std.debug.print("\n", .{});

    // 3. Range filtering
    std.debug.print("3. Range Filtering\n\n", .{});

    const all_addresses = [_]Address{
        Address.fromU256(10),
        Address.fromU256(50),
        Address.fromU256(100),
        Address.fromU256(200),
        Address.fromU256(500),
    };

    const min = Address.fromU256(50);
    const max = Address.fromU256(200);

    const min_hex = Address.toHex(min);
    const max_hex = Address.toHex(max);
    std.debug.print("Range: {s} to {s}\n\n", .{ &min_hex, &max_hex });

    var in_range = std.ArrayList(Address).init(a);
    defer in_range.deinit();

    for (all_addresses) |addr| {
        if (!Address.lessThan(addr, min) and !Address.greaterThan(addr, max)) {
            try in_range.append(addr);
        }
    }

    std.debug.print("Addresses in range:\n", .{});
    for (in_range.items) |addr| {
        const hex = Address.toHex(addr);
        std.debug.print("  {s}\n", .{&hex});
    }
    std.debug.print("\n", .{});

    // 4. Deduplication
    std.debug.print("4. Deduplication\n\n", .{});

    const with_duplicates = [_]Address{
        try Address.fromHex("0x742d35Cc6634C0532925a3b844Bc9e7595f51e3e"),
        try Address.fromHex("0xa0Cf798816D4b9b9866b5330EEa46a18382f251e"),
        try Address.fromHex("0x742d35cc6634c0532925a3b844bc9e7595f51e3e"), // Duplicate (different case)
        try Address.fromHex("0xa0Cf798816D4b9b9866b5330EEa46a18382f251e"), // Duplicate (exact)
        try Address.fromHex("0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266"),
    };

    std.debug.print("Original ({} addresses):\n", .{with_duplicates.len});
    for (with_duplicates) |addr| {
        const hex = Address.toChecksummed(addr);
        std.debug.print("  {s}\n", .{&hex});
    }
    std.debug.print("\n", .{});

    var unique = std.ArrayList(Address).init(a);
    defer unique.deinit();

    for (with_duplicates) |addr| {
        var found = false;
        for (unique.items) |existing| {
            if (Address.equals(addr, existing)) {
                found = true;
                break;
            }
        }
        if (!found) {
            try unique.append(addr);
        }
    }

    std.debug.print("Deduplicated ({} addresses):\n", .{unique.items.len});
    for (unique.items) |addr| {
        const hex = Address.toChecksummed(addr);
        std.debug.print("  {s}\n", .{&hex});
    }
    std.debug.print("\n", .{});

    // 5. Finding specific addresses
    std.debug.print("5. Finding Specific Addresses\n\n", .{});

    const address_list = [_]Address{
        try Address.fromHex("0x742d35Cc6634C0532925a3b844Bc9e7595f51e3e"),
        try Address.fromHex("0xa0Cf798816D4b9b9866b5330EEa46a18382f251e"),
        try Address.fromHex("0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266"),
    };

    const target = try Address.fromHex("0xa0Cf798816D4b9b9866b5330EEa46a18382f251e");

    // Find index
    var index: ?usize = null;
    for (address_list, 0..) |addr, i| {
        if (Address.equals(addr, target)) {
            index = i;
            break;
        }
    }

    const target_hex = Address.toChecksummed(target);
    std.debug.print("Target: {s}\n", .{&target_hex});
    if (index) |idx| {
        std.debug.print("Found at index: {}\n", .{idx});
    } else {
        std.debug.print("Not found\n", .{});
    }
    std.debug.print("\n", .{});

    // 6. Grouping by prefix
    std.debug.print("6. Grouping by Prefix\n\n", .{});

    const mixed_addresses = [_]Address{
        try Address.fromHex("0x0000000000000000000000000000000000000001"),
        try Address.fromHex("0x0000000000000000000000000000000000000002"),
        try Address.fromHex("0x1111111111111111111111111111111111111111"),
        try Address.fromHex("0x1111111111111111111111111111111111111112"),
        try Address.fromHex("0xffffffffffffffffffffffffffffffffffffffff"),
    };

    var groups = std.AutoHashMap(u8, std.ArrayList(Address)).init(a);
    defer {
        var it = groups.iterator();
        while (it.next()) |entry| {
            entry.value_ptr.deinit();
        }
        groups.deinit();
    }

    for (mixed_addresses) |addr| {
        const first_byte = addr[0];
        const result = try groups.getOrPut(first_byte);
        if (!result.found_existing) {
            result.value_ptr.* = std.ArrayList(Address).init(a);
        }
        try result.value_ptr.append(addr);
    }

    std.debug.print("Grouped by first byte:\n", .{});
    var it = groups.iterator();
    while (it.next()) |entry| {
        std.debug.print("  0x{x:0>2}: {} addresses\n", .{ entry.key_ptr.*, entry.value_ptr.items.len });
        for (entry.value_ptr.items) |addr| {
            const hex = Address.toChecksummed(addr);
            std.debug.print("    {s}\n", .{&hex});
        }
    }
    std.debug.print("\n", .{});

    // 7. Top N addresses
    std.debug.print("7. Top N Addresses (Largest Values)\n\n", .{});

    var many_addresses = [_]Address{
        Address.fromU256(100),
        Address.fromU256(500),
        Address.fromU256(50),
        Address.fromU256(1000),
        Address.fromU256(200),
        Address.fromU256(750),
    };

    // Sort descending
    std.mem.sort(Address, &many_addresses, {}, struct {
        fn lessThan(_: void, lhs: Address, rhs: Address) bool {
            return Address.compare(lhs, rhs) > 0; // Reversed for descending
        }
    }.lessThan);

    const top_n = 3;
    std.debug.print("Top {} addresses:\n", .{top_n});
    for (many_addresses[0..top_n], 0..) |addr, i| {
        const hex = Address.toHex(addr);
        const value = Address.toU256(addr);
        std.debug.print("  {}. {s} ({})\n", .{ i + 1, &hex, value });
    }
    std.debug.print("\n", .{});

    // 8. Performance note
    std.debug.print("8. Performance Note\n\n", .{});
    std.debug.print("For optimal performance when working with collections:\n", .{});
    std.debug.print("- Use HashMap with hex string as key for lookups\n", .{});
    std.debug.print("- Use Address.compare() for sorting\n", .{});
    std.debug.print("- Use Address.equals() for exact matching\n", .{});
    std.debug.print("- Addresses are [20]u8 - lightweight and stack-allocated\n", .{});
}
